man/mptcpize.8.in
.\" SPDX-License-Identifier: BSD-3-Clause
.\"
.\" Copyright (c) 2021, Intel Corporation

.\" Process this file with
.\" groff -man -Tascii mptcpize.8
.\"
.TH MPTCPIZE 8 "2021-09-23" "Multipath TCP Daemon" "System Management Commands"
.SH NAME
mptcpize \- enable MPTCP on existing services
.SH SYNOPSIS
.SY mptcpize
.OP \-?V
.OP \-\-help
.OP \-\-usage
.OP \-\-version
.B COMMAND
.YS

.SH DESCRIPTION
.B mptcpize
is a program that enables multipath TCP on existing legacy services,
where
.B COMMAND
is one of the following:

.SS
.BI run\ [ -d ]\  prog \ [ args ]
Run target program with the specified command line arguments, forcing
MPTCP socket usage instead of TCP.  If the
.B -d
argument is provided, dump messages on
.B stderr
when a TCP socket is forced to use MPTCP.

.SS
.BI enable\  unit
Update the systemd
.I unit
file, forcing the given service to run under the above launcher.

.SS
.BI disable\  unit
Update the systemd
.I unit
file, removing the above launcher.


.SH OPTIONS
.B mptcpize
accepts the following command line options:

.TP
.BR \-? , \-\-help
display
.B mptcpize
help information

.TP
.B \-\-usage
display brief
.B mptcpize
usage information

.TP
.BR \-V , \-\-version
display
.B mptcpize
version information

.SH REPORTING BUGS
Report bugs to
.MT @PACKAGE_BUGREPORT@
.ME .

.SH SEE ALSO
ip-mptcp(8), mptcpd(8)

.\" Local Variables:
.\" mode: nroff
.\" End:


NEWS
24 September 2021 - mptcpd 0.8

- The mptcpd library API now splits path management operations
  specific to the in-kernel path manager found in the upstream Linux
  kernel into a separate "mptcpd_kpm" API namespace to differentiate
  ADD_ADDR and REMOVE_ADDR related operations between the in-kernel
  and user space cases.  The operations for the two are different and
  have different use cases.

- Two new mptcpd command line arguments were added: "--addr-flags" and
  "--notify-flags", along with the equivalent settings in the mptcpd
  system configuration file (e.g., /etc/mptcpd/mptcpd.conf).
  "addr-flags" are used when announcing an IP address.  Similar flags
  are used by the "ip mptcp" sub-command.  See the ip-mptcp(8) and man
  page for further details on these address related flags.
  "notify-flags" provides for further control over how plugins are
  notified of changes to local IP addresses.  See the mptcpd(8) man
  page shipped with this mptcpd release for further details.

- Improve the mptcpd "addr_adv" plugin by making it set suitable MPTCP
  resource limits in the kernel, such as expanding the maximum number
  of subflows to allow subflows associated with advertised IP
  addresses to be created.

- The mptcpd plugin directory name in the mptcpd system configuration
  file may now be left empty to improve "multilib" support found in
  some Linux distributions.  A compile-time default will be used if no
  plugin directory name is found in the mptcpd system configuration
  file or the mptcpd command line options.

- Mptcpd command line options now properly take precedence over the
  corresponding mptcpd system configuration setting.  This addresses
  an issue that prevented mptcpd from starting if a required setting
  was not specified in the mptcpd system configuration file even
  though the same setting was configured through the mptcpd command
  line.

- A new "mptcpize" program was added that allows legacy TCP-only
  applications to transparently use MPTCP by either of the following
  approaches:
    - leverage library inter-positioning to transparently replace TCP
      socket calls with their MPTCP counterparts.
    - enable or disable TCP to MPTCP socket conversion through an
      existing systemd unit file.

- Build regressions against ELL versions 0.31 and 0.33 were
  corrected.

13 April 2021 - mptcpd 0.7

- MPTCP path management generic netlink events recently added to the
  upstream Linux kernel (v5.12) are now supported, and propagated to
  mptcpd plugins.  The same API is also found in the multipath-tcp.org
  kernel.  Differences between the two kernels are transparent to
  mptcpd plugins.

- Mptcpd now supports the upstream Linux kernel server-oriented
  MPTCP_PM_CMD_SET_FLAGS generic netlink command through the new
  mptcpd_pm_set_flags() function.  It allows the user to set MPTCP
  flags for a specific local IP address.

- An incorrect buffer size calculation when calling
  mptcpd_pm_add_addr() when using an IPv6 addresses on platforms
  running the upstream Linux kernel was corrected.

- An "operation not supported" error that occurred when attempting to
  retrieve MPTCP resource limits through the mptcpd_pm_get_limits()
  function was fixed.

- A network interface lookup problem in the "sspi" plugin was
  corrected.

- Building mptcpd against ELL 0.39 is now supported.

25 January 2021 - mptcpd 0.6

- Mptcpd now supports versions of the Embedded Linux Library (ELL)
  greater than 0.33.

- Plugins should use the new MPTCPD_PLUGIN_DEFINE() preprocessor macro
  instead of L_PLUGIN_DEFINE().

- A pointer to the mptcpd path manager object, i.e. struct mptcpd *pm,
  is now passed to the plugin init and exit functions.  This allows
  plugins to potentially perform mptcpd path manager related
  operations during initialization and finalization.

- Support for the MPTCP netlink path manager in the upstream Linux
  kernel is now available.  A new set of path management command
  functions corresponding to those available in the kernel netlink
  path management API has been added to the `<mptcpd/path_manager.h>'
  header.  The new functions allow plugins to retrieve IP address
  information, flush addresses, and modify MPTCP resource limits.

- The mptcpd_pm_add_addr() (formerly mptcpd_pm_send_addr()) and
  mptcpd_pm_remove_addr() function parameters have been modified in
  order to support both the upstream and multipath-tcp.org kernels.

- Mptcpd path management command functions declared in
  `<mptcpd/path_manager.h>' now return zero on success and -1 or an
  errno on failure instead of a bool.

- A MPTCP address ID manager "mptcpd_idm" interface was introduced
  that mptcpd plugins may leverage to map an IP address to a MPTCP
  address ID, as well as to track used and unused IDs.  The interface
  is defined in the new `<mptcpd/id_manager.h>' header.

- A new address advertising plugin, "addr_adv", has been added.  It
  simply triggers a MPTCP ADD_ADDR when a new IP address is detected
  by the mptcpd network monitor.  Similarly, a MPTCP REMOVE_ADDR is
  triggered when an IP address is no longer available.

- MPTCP netlink command error message logging was improved to be more
  descriptive when possible.

12 June 2020 - mptcpd 0.5

- Test plugins are no longer installed when running "make install".  A
  side benefit of the changes is that test plugins only build when the
  test suite is built via "make check" or "make distcheck".  Libtool
  relink warnings that occurred when building the tests have also been
  addressed.

- VPATH builds where the build directory and source directory
  differ (e.g. "mkdir build; cd build; ../configure; make") now
  succeed.

21 May 2020 - mptcpd 0.4

- Plugins may now inform peers that an address is no longer being
  advertised through the new mptcpd_pm_remove_addr() function.  This
  corresponds to the MPTCP protocol REMOVE_ADDR option.

- Network monitoring related events are now propagated to plugins.

src/mptcpize.c
// SPDX-License-Identifier: BSD-3-Clause
/**
 * @file mptcpize.c
 *
 * @brief enable mptcp on existing services.
 *
 * Copyright (c) 2021, Red Hat, Inc.
 */

#define  _GNU_SOURCE

#include <linux/limits.h>

#include <sys/types.h>
#include <sys/stat.h>

#include <argp.h>
#include <dlfcn.h>
#include <errno.h>
#include <error.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#ifdef HAVE_CONFIG_H
# include <mptcpd/private/config.h>
#endif

#define SYSTEMD_ENV_VAR		"Environment="
#define SYSTEMD_UNIT_VAR	"FragmentPath="
#define SYSTEMD_SERVICE_TAG	"[Service]"
#define SYSTEMCTL_SHOW		"systemctl show -p FragmentPath "
#define PRELOAD_VAR		"LD_PRELOAD="
#define MPTCPWRAP_ENV		"LD_PRELOAD="PKGLIBDIR"/libmptcpwrap.so.0.0."LIBREVISION

/* Program documentation. */
static char args_doc[] = "CMD";

static char doc[] =
        "mptcpize - a tool to enable MPTCP usage on unmodified legacy services\v"
        "Available CMDs:\n"
        "\trun [-d] prog [<args>]    Run target program with specified\n"
        "\t                          arguments, forcing MPTCP socket usage\n"
        "\t                          instead of TCP.  If the '-d' argument\n"
        "\t                          is provided, dump messages on stderr\n"
        "\t                          when a TCP socket is forced to MPTCP.\n\n"
        "\tenable <unit>             Update the systemd <unit>, forcing\n"
        "\t                          the given service to run under the\n"
        "\t                          above launcher.\n\n"
        "\tdisable <unit>            Update the systemd <unit>, removing\n"
        "\t                          the above launcher.\n";

static struct argp const argp = { 0, 0, args_doc, doc, 0, 0, 0 };

static void help(void)
{
	argp_help(&argp, stderr, ARGP_HELP_STD_HELP, "mptcpize");
}

static int run(int argc, char *av[])
{
	int i, nr = 0, debug = 0;
	char **envp, **argv;

	if (argc > 0 && strcmp(av[0], "-d") == 0) {
		debug = 1;
		argc--;
		av++;
	}

	if (argc < 1) {
		fprintf(stderr, "missing command argument\n");
		help();
		return -1;
	}

	// build environment, copying the current one ...
	while (environ[nr])
		nr++;
	envp = calloc(nr + 3, sizeof(char *));
	if (!envp)
		error(1, errno, "can't allocate env list");

	// ... filtering out any 'LD_PRELOAD' ...
	nr = 0;
	i = 0;
	while (environ[nr]) {
		if (strncmp(environ[nr], PRELOAD_VAR,
			    strlen(PRELOAD_VAR)) != 0) {
			envp[i] = environ[nr];
			i++;
		}
		nr++;
	}

	// ... appending the mptcpwrap preload...
	envp[i++] = MPTCPWRAP_ENV;

	// ... and enable dbg if needed
	if (debug)
		envp[i++] = "MPTCPWRAP_DEBUG=1";

	// build the NULL terminated arg list
	argv = calloc(argc + 1, sizeof(char *));
	if (!argv)
		error(1, errno, "can't allocate argument list");

	memcpy(argv, av, argc * sizeof(char*));
	return execvpe(argv[0], argv, envp);
}

static char *locate_unit(const char *name)
{
	char *cmd, *line = NULL;
	FILE *systemctl;
	size_t len = 0;
	ssize_t read;

	/* check for existing unit file */
	if (access(name, R_OK) == 0)
		return strdup(name);

	/* this is supposed to be an unit name */
	len = strlen(name) + 1 + strlen(SYSTEMCTL_SHOW);
	cmd = malloc(len);
	if (!cmd)
		error(1, 0, "can't allocate systemctl command string");

	sprintf(cmd, SYSTEMCTL_SHOW"%s", name);
	systemctl = popen(cmd, "r");
	if (!systemctl)
		error(1, errno, "can't execute %s", cmd);

	free(cmd);
	while ((read = getline(&line, &len, systemctl)) != -1) {
		if (strncmp(line, SYSTEMD_UNIT_VAR, strlen(SYSTEMD_UNIT_VAR)) == 0) {
			char *ret = strdup(&line[strlen(SYSTEMD_UNIT_VAR)]);
			if (!ret)
				error(1, errno, "failed to duplicate string");

			// trim trailing newline, if any
			len = strlen(ret);
			if (len > 0 && ret[len - 1] == '\n')
				ret[--len] = 0;
			if (len == 0)
				error(1, 0, "can't find unit file for service %s", name);
			free(line);
			pclose(systemctl);
			return ret;
		}
	}

	error(1, 0, "can't find FragmentPath attribute for unit %s", name);

	// never reached: just silence gcc
	return NULL;
}

static int unit_update(int argc, char *argv[], int enable)
{
	char *unit, *line = NULL;
	int append_env = enable;
	char dst_path[PATH_MAX];
	size_t len = 0;
	ssize_t read;
	FILE *src;
	int dst;

	if (argc < 1) {
		fprintf(stderr, "missing unit argument\n");
		help();
		return -1;
	}

	unit = locate_unit(argv[0]);
	src = fopen(unit, "r");
	if (!src)
		error(1, errno, "can't open file %s", unit);

	strcpy(dst_path, "/tmp/unit_XXXXXX");
	dst = mkstemp(dst_path);
	if (dst < 0)
		error(1, errno, "can't create tmp file");

	// reset any prior error, to allow later check on errno
	errno = 0;

	/**
	 * systemd does not allow Environment property update via a command,
	 * we have to copy the existing unit file to a tmp one
	 */
	while ((read = getline(&line, &len, src)) != -1) {
		int is_env = strncmp(line, SYSTEMD_ENV_VAR, strlen(SYSTEMD_ENV_VAR)) == 0;

		if (!is_env) {
			if (write(dst, line, read) < 0)
				error(1, errno, "can't write '%s' into %s", line, dst_path);
		}

		if (append_env &&
		    (is_env || strncmp(line, SYSTEMD_SERVICE_TAG, strlen(SYSTEMD_SERVICE_TAG)) == 0)) {
			if (dprintf(dst, "%s%s\n", SYSTEMD_ENV_VAR, MPTCPWRAP_ENV) < 0)
				error(1, errno, "can't write to env string into %s", dst_path);
			append_env = 0;
		}
	}
	if (errno != 0)
		error(1, errno, "can't read from %s", unit);
	free(line);
	fclose(src);
	close(dst);

	if (rename(dst_path, unit) < 0)
		error(1, errno, "can't rename %s to %s", dst_path, unit);

	if (system("systemctl daemon-reload") != 0)
		error(1, errno, "can't reload unit, manual 'systemctl daemon-reload' is required");

	printf("mptcp successfully %s on unit %s\n",
	       enable ? "enabled" : "disabled", unit);
	free(unit);
	return 0;
}

static int enable(int argc, char *argv[])
{
	return unit_update(argc, argv, 1);
}

static int disable(int argc, char *argv[])
{
	return unit_update(argc, argv, 0);
}

int main(int argc, char *argv[])
{
	int idx;

	argp_program_version = "mptcpize "VERSION;
	argp_program_bug_address = "<" PACKAGE_BUGREPORT ">";
	if (argp_parse(&argp, argc, argv, ARGP_IN_ORDER, &idx, 0) < 0)
		error(1, errno, "can't parse arguments");

	argc -= idx;
	argv += idx;
	while (argc > 0) {
		if (strcmp(argv[0], "run") == 0)
			return run(--argc, ++argv);
		else if (strcmp(argv[0], "enable") == 0)
			return enable(--argc, ++argv);
		else if (strcmp(argv[0], "disable") == 0)
			return disable(--argc, ++argv);
		else if (strcmp(argv[0], "help") == 0) {
			help();
			return 0;
		} else {
			fprintf(stderr, "unknown arg %s\n", argv[0]);
			return -1;
		}
	}
	help();
	return 0;
}

/*
  Local Variables:
  c-file-style: "linux"
  End:
*/
